### PR TITLE
feat: Enable signing with pre-computed hash

### PR DESCRIPTION
**Pull request by laurentsimon**
closes https://github.com/sigstore/sigstore-python/issues/666

Let me know if you'd like me to add a dedicated unit test for this new option.

Context: I'm working on adding a new hash type to rekor to support hashing directories. I will need to follow up to add support for it in https://github.com/trailofbits/sigstore-rekor-types. 

Test example with pre-computed sha256:
```python
from sigstore import hashes as sigstore_hashes
from sigstore_protobuf_specs.dev.sigstore.common.v1 import HashAlgorithm

with file.open(mode="rb", buffering=0) as io:
    digest = hashlib.sha256(io.read())
    hashed = sigstore_hashes.Hashed(digest=digest, algorithm=HashAlgorithm.SHA2_256)
    result = signer.sign(hashed)
```
#### Summary

This PR enables passing a pre-hashed value instead of the content to be hashed.

### CODE REVIEW

File: sigstore/sign.py
Comment (laurentsimon) [inline sigstore/sign.py:175]: ignore those links

File: sigstore/sign.py
Comment (laurentsimon) [inline sigstore/sign.py:193]: ignore the print

File: sigstore/sign.py
Comment (jku) [inline sigstore/sign.py:191]: I assume you mean Prehashed here?

File: sigstore/sign.py
Comment (jku) [inline sigstore/sign.py:415]: Can you talk me through this (possibly I just don't understand the protobuf details):
* we're creating protobuf message HashOutput where algorithm is a string (that comes from cryptography?)
* but the protobuf HashAlgorithm is actually an enum 
* HashAlgorithm has a single documented value (SHA2_256)

Comment (laurentsimon) [reply sigstore/sign.py]: HashAlgorithm comes from Sigstore `sigstore_protobuf_specs.dev.sigstore.common.v1` which comes from `https://github.com/sigstore/protobuf-specs/blob/d96fa4407cb4489233f133ab036c0ed98c2756a4/protos/sigstore_common.proto#L39`. The value is a string, ifaict.

File: sigstore/sign.py
Comment (laurentsimon) [inline sigstore/sign.py:176]: Note: I considered using `input_ IO[bytes] | bytes` where `bytes` type would be used with `Prehashed` but I think it further complicates the API. So I've kept `IO[bytes]` instead. Please advise.

**Comment by laurentsimon** [comment]
tests seem to be failing because the bundle uses a different name for the hash algo (`SHA2_256`) instead of `SHA256` for the standard crypto lib. I'll fix that while you review the PR.

File: sigstore/sign.py
Comment (jku) [reply sigstore/sign.py]: I don't think it works that way. The python bindings might not complain when that r.uns it but the protobuf serialization likely won't work.

But I suppose if you were asking for high level feedback, the real questions related to this are: 
* how is this going to work in bundle that is an interop format? 
* should the bundle even try to document the used algorithm (since the verifier I guess is anyway expected to know the magic precalculation method that is required to actually verify) -- does the protobuf need a "custom" algorithm value?

**Comment by jku** [comment]
If you run `make reformat` you'll start to get useful lint results as well

File: sigstore/_utils.py
Comment (laurentsimon) [inline sigstore/_utils.py:144]: naive attempt to fix the conformance tests without running them locally. Still complaining that the enum number is invalid

File: sigstore/sign.py
Comment (laurentsimon) [reply sigstore/sign.py]: re: yeah you're right about types I double checked. I replied to your question about bundle in the other comment. TL;DR: We're adding a hash in https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_common.proto#L39 which will rekor will support for hashedrekord. This PR _only_ makes it possible to pre-compute _a_ hash supported by rekor. Right it only supports sha256 because rekor only supports that hash.

Let me know if I'm misunderstanding something. Thanks

Comment (laurentsimon) [reply sigstore/sign.py]: Please also see https://github.com/trailofbits/sigstore-rekor-types/issues/22

**Comment by laurentsimon** [comment]
> If you run `make reformat` you'll start to get useful lint results as well

I've run these commands and it works locally. However in CI it's failing. Any ideas?

File: sigstore/verify/models.py
Comment (laurentsimon) [inline sigstore/verify/models.py:182]: This seems like a breaking change :(

**Comment by laurentsimon** [comment]
I've made the changes, PTAL

File: sigstore/verify/models.py
Comment (woodruffw) [reply sigstore/verify/models.py:182]: Yep, it is. That's fine IMO since we're on a breakage path for 3.0 anyways, and #804 will also break compatibility here.

File: sigstore/_utils.py
Comment (woodruffw) [inline sigstore/_utils.py:149]: Extreme nitpick, but IMO we can drop this `NOTE`: the `sha256_streaming` call's parameter hint will keep this checked for us.

**Comment by woodruffw** [comment]
> I've run these commands and it works locally. However in CI it's failing. Any ideas?

Hmm, this might be an outdated local toolchain. What does `ruff --version` say?

File: sigstore/hashes.py
Comment (woodruffw) [inline sigstore/hashes.py:1]: Thinking out loud: this is a pretty small API surface for its own module. Rather than committing to a new API path like this (`sigstore.hashes`), maybe we could do one of the following:

* Turn this into `sigstore._hashes`, and re-export `Hashed` from `sigstore.sign`
* Remove this entirely, and inline it into `sigstore.sign`

Thoughts on either of these?

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:325]: NB: Both of these fields will disappear with #804 and we intend to remove `SigningResult` entirely with #854, so we should probably avoid more fields here. Given that `HashAlgorithm` will currently always be invariant, maybe we can just hardcode it with a `NOTE` for now?

File: sigstore/hashes.py
Comment (laurentsimon) [reply sigstore/hashes.py:1]: does it matter that that the `verify/*` will be needing it? If not, I'm fine with any of the above.

File: sigstore/sign.py
Comment (laurentsimon) [reply sigstore/sign.py:325]: as soon as we update rekor (probably next week), it will no longer be an invariant, ie there wil be 2 supported hashes (sha256 and some dirhash hash)

Comment (woodruffw) [reply sigstore/sign.py:325]: Gotcha -- in that case, we should probably bump up the priority on #854, since `SigningResult` has outlived its utility in the world of Sigstore bundles 🙂

**Comment by woodruffw** [comment]
> I updated to ruff `0.1.13`. I managed to locally fix the files. CI still complaining.

Gotcha, sorry for the mess here -- I'll do some debugging on the CI today. It looks like `lint` is still failing, although this time because we have public APIs without docstrings.

File: sigstore/sign.py
Comment (laurentsimon) [reply sigstore/sign.py:325]: SGTM. So you'd land https://github.com/sigstore/sigstore-python/issues/854 then this PR I suppose? Or you don't really care?

Comment (woodruffw) [reply sigstore/sign.py:325]: Tangent, but is dirhash documented anywhere? I see there's a dirhash standard on GitHub but I wasn't sure if it was the same thing.

(Part of what I'm wondering is what the mapping between a dirhash and a standard algorithm is -- how do we form a `Prehashed` for them?)

File: sigstore/hashes.py
Comment (woodruffw) [reply sigstore/hashes.py:1]: Ah yeah, good point. Given that, I think this is fine as its own module.

Comment (laurentsimon) [reply sigstore/hashes.py:1]: Alright. Keeping as-is then. Thanks

**Comment by laurentsimon** [comment]
@woodruffw out of curiosity, what is the timeline for v3 release https://github.com/sigstore/sigstore-python/milestone/5?

**Comment by woodruffw** [comment]
> @woodruffw out of curiosity, what is the timeline for v3 release https://github.com/sigstore/sigstore-python/milestone/5?

I don't think we have a fixed timeline yet, it's mostly just a matter of when DSSE support and this get merged in, along with a few CLI deprecations. 

(To speed things along, we could definitely do one or more v3 prereleases, if you're looking to consume the APIs downstream.)

File: sigstore/sign.py
Comment (woodruffw) [reply sigstore/sign.py:325]: That matches my understanding, thanks! We should probably make sure that subclassing `HashAlgorithm` like that for external hashes is okay in terms of PyCA's APIs, but otherwise that seems reasonable to me.

**Comment by laurentsimon** [comment]
> > @woodruffw out of curiosity, what is the timeline for v3 release https://github.com/sigstore/sigstore-python/milestone/5?
> 
> I don't think we have a fixed timeline yet, it's mostly just a matter of when DSSE support and this get merged in, along with a few CLI deprecations.
> 
> (To speed things along, we could definitely do one or more v3 prereleases, if you're looking to consume the APIs downstream.)

That'd be nice yes! I'd be interested in a prelrelease with this PR merged in, so we can start testing e2e on our side. What would be a timeline for it?

**Comment by woodruffw** [comment]
> That'd be nice yes! I'd be interested in a prelrelease with this PR merged in, so we can start testing e2e on our side. What would be a timeline for it?

We could make that pre-release as soon as this gets merged in 🙂 

(We won't make any semver guarantees around pre-releases, but I think that's already given.)

**Comment by laurentsimon** [comment]
> > That'd be nice yes! I'd be interested in a prelrelease with this PR merged in, so we can start testing e2e on our side. What would be a timeline for it?
> 
> We could make that pre-release as soon as this gets merged in 🙂
> 
> (We won't make any semver guarantees around pre-releases, but I think that's already given.)

I'd probably just ask that the rekor changes are in too. But I'll let you know once this is done.

**Comment by laurentsimon** [comment]
I'll wait for https://github.com/sigstore/sigstore-python/pull/862 to land and update this PR. Anything else I should address in the PR?

**Comment by woodruffw** [comment]
> I'll wait for #862 to land and update this PR. Anything else I should address in the PR?

Nope, I think you're good. Thanks a ton for your work here!

**Comment by jku** [comment]
Looks good (didn't try it yet though). 

I just have the one comment, same as before: using cryptography to lookup algorithm names for some other system smells a bit.

This is about hazmat_algorithm():
* the method is only used outside of the class to get the algorithm name (`as_prehashed()` could easily avoid using it)
* the algo name is supposed to be compatible with _other sigstore components, not cryptography_: we can't expect the hazmat algorithms to use the same strings that sigstore uses. 

I'm not 100% sure if this is complete but... does this do what you need: 
```
def as_prehashed(self) -> Prehashed:
    if self.algorithm == HashAlgorithm.SHA2_256:
        return Prehashed(hashes.SHA256())
    raise ...

def as_hashedrekord_algorithm() -> hashedrekord.Algorithm
    if self.algorithm == HashAlgorithm.SHA2_256:
            return hashedrekord.Algorithm.SHA256
    raise ...
```

**Comment by woodruffw** [comment]
#862 is landed, so this needs a deconflict. I'm also planning to merge #804 soon, so you _may_ want to wait until that's landed before deconflicting (since it touches many of the same lines).

**Comment by laurentsimon** [comment]
> #862 is landed, so this needs a deconflict. I'm also planning to merge #804 soon, so you _may_ want to wait until that's landed before deconflicting (since it touches many of the same lines).

OK I'll wait for #804 to land then.

**Comment by jku** [comment]
#804 is in main now

**Comment by laurentsimon** [comment]
I've rebased. However I was not able to tests locally, keeps getting this error:

`spec=rekor_types.hashedrekord.HashedrekordV001Schema(
         ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'rekor_types' has no attribute 'hashedrekord'. Did you mean: 'Hashedrekord'?`

**Comment by jku** [comment]
> How do you typically test?

the makefile tries to manage a virtual env for testing:
* this gets used by `make test` or can be generated by `make dev`
* once the env exists exists you can enter it for manual testing with `source env/bin/activate`

FYI, the dependency version pinning is a bit strange:
* pyproject.toml documents the actual **required** runtime versions 
* install/requirements.txt documents pinned versions -- but this is **only** updated at release time

For sigstore-rekor-types: you want 0.0.12 as documented by pyproject.toml. The virtual environment should already have this version.

**Comment by laurentsimon** [comment]
> > How do you typically test?
> 
> the makefile tries to manage a virtual env for testing:
> 
> * this gets used by `make test` or can be manually generated by `make dev`
> * once the env exists exists you can enter it for manual testing with `source env/bin/activate`
> 
> FYI, the dependency version pinning is a bit strange:
> 
> * pyproject.toml documents the actual **required** runtime versions
> * install/requirements.txt documents pinned versions -- but this is **only** updated at release time (I think this is a bit confusing and I'm not sure what the purpose is)
> 
> For sigstore-rekor-types: you want 0.0.12 as documented by pyproject.toml. The virtual environment should already have this version.

Thank you, that was super helpful. I was able to make the changes and the conformance tests are now passing. PTAL.

**Review by woodruffw — COMMENTED** [review]
This looks great to me @laurentsimon! One last question, and then I think we're good to merge here.

File: sigstore/hashes.py
Comment (laurentsimon) [reply sigstore/hashes.py:45]: +1, they do not. Let's move them to private. I meant to ask and forgot. Alternatively I make the entire file private `_hash.py`. What do you prefer?

Comment (woodruffw) [reply sigstore/hashes.py:45]: Given that `Hashed` will be in the publish signing API, I think we can leave `hashes.py` public!

Comment (laurentsimon) [reply sigstore/hashes.py]: > Given that Hashed will be in the public signing API, I think we can leave hashes.py public!

what was I thinking... changes made. Thanks

File: sigstore/_cli.py
Comment (woodruffw) [inline sigstore/_cli.py:694]: Can we move these imports to the top level? Inline imports like this tend to be a lot harder to test/easy to introduce circularities with 🙂

File: sigstore/_cli.py
Comment (woodruffw) [inline sigstore/_cli.py:703]: Actually, looking at this more, why are we doing the pre-hash here? I think `sign(...)` should probably continue to take `IO[bytes]` in this case, right?

File: sigstore/_cli.py
Comment (laurentsimon) [reply sigstore/_cli.py]: accidental commit that was only used for testing. Removed.

**Comment by woodruffw** [comment]
LGTM, although lint is still failing. @laurentsimon are you able to diagnose that? Otherwise I can try and do it, if you give me push access to your branch.

**Comment by laurentsimon** [comment]
> LGTM, although lint is still failing. @laurentsimon are you able to diagnose that? Otherwise I can try and do it, if you give me push access to your branch.

linter passes locally, so I think you're in a better position to debug that part :)
You have write access to the PR branch, I think

**Comment by woodruffw** [comment]
Fixed up the CI (sorry for the pain here, I need to figure out what's going on).

This LGTM, but @jku or another maintainer will need to approve now that I'm the last pusher. Given that, I'll spend some time writing some tests here.

File: test/unit/test_sign.py
Comment (woodruffw) [inline test/unit/test_sign.py:148]: I've added this to confirm that the round-trip works as expected: we sign over a `Hashed(...)` and then verify over the pre-image.

**Review by woodruffw — APPROVED** [review]

**Review by jku — APPROVED** [review]
Looks good!

**Comment by woodruffw** [comment]
Thanks a ton for your hard work here @laurentsimon!

**Pull request merged**
